crates/task-impls/src/upgrade.rs

```diff
@@ -0,0 +1,255 @@
+
+    /// Reference to consensus. Leader will require a read lock on this.
+    pub consensus: Arc<RwLock<Consensus<TYPES>>>,
+
+    /// Membership for Quorum Certs/votes
+    pub quorum_membership: Arc<TYPES::Membership>,
+    /// Network for all nodes
+    pub quorum_network: Arc<I::QuorumNetwork>,
+
+    /// Whether we should vote affirmatively on a given upgrade proposal (true) or not (false)
+    pub should_vote: fn(UpgradeProposalData<TYPES>) -> bool,
+
+    /// The current vote collection task, if there is one.
+    pub vote_collector:
+        RwLock<VoteCollectorOption<TYPES, UpgradeVote<TYPES>, UpgradeCertificate<TYPES>>>,
+
+    /// Global events stream to publish events
+    pub event_stream: ChannelStream<HotShotEvent<TYPES>>,
+
+    /// This Nodes public key
+    pub public_key: TYPES::SignatureKey,
+
+    /// This Nodes private key
+    pub private_key: <TYPES::SignatureKey as SignatureKey>::PrivateKey,
+
+    /// This state's ID
+    pub id: u64,
+}
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> + 'static>
+    UpgradeTaskState<TYPES, I, A>
+{
+    /// main task event handler
+    #[instrument(skip_all, fields(id = self.id, view = *self.cur_view), name = "Upgrade Task", level = "error")]
+    pub async fn handle_event(
+        &mut self,
+        event: HotShotEvent<TYPES>,
+    ) -> Option<HotShotTaskCompleted> {
+        match event {
+            HotShotEvent::UpgradeProposalRecv(proposal, sender) => {
+                let should_vote = self.should_vote;
+                // If the proposal does not match our upgrade target, we immediately exit.
+                if should_vote(proposal.data.upgrade_proposal.clone()) {
+                    warn!(
+                        "Received unexpected upgrade proposal:\n{:?}",
+                        proposal.data
+                    );
+                    return None;
+                }
+
+                // If we have an upgrade target, we validate that the proposal is relevant for the current view.
+
+                debug!(
+                    "Upgrade proposal received for view: {:?}",
+                    proposal.data.get_view_number()
+                );
+                // NOTE: Assuming that the next view leader is the one who sends an upgrade proposal for this view
+                let view = proposal.data.get_view_number();
+
+                // Allow an upgrade proposal that is one view older, in case we have voted on a quorum
+                // proposal and updated the view.
+                // `self.cur_view` should be at least 1 since there is a view change before getting
+                // the `UpgradeProposalRecv` event. Otherewise, the view number subtraction below will
+                // cause an overflow error.
+                // TODO Come back to this - we probably don't need this, but we should also never receive a UpgradeCertificate where this fails, investigate block ready so it doesn't make one for the genesis block
```

Comment:
I think we need to keep the check in, a malicious party could send upgrades for old views, maybe to make use vote.  I think it's good to check for recency

Reply:
If you're talking about the genesis check, then I agree we can maybe revisit that par